gemrb/core/Projectile.cpp

```diff
@@ -1196,48 +1200,65 @@ void Projectile::SecondaryTarget()
 
 	//In case of utter failure, apply a spell of the same name on the caster
 	//this feature is used by SCHARGE, PRTL_OP and PRTL_CL in the HoW pack
-	if(fail) {
+	if (fail) {
 		ApplyDefault();
 	}
 }
 
-int Projectile::Update()
-{
-	//if reached target explode
-	//if target doesn't exist expire
-	if (phase == P_EXPIRED) {
-		return 0;
+void Projectile::Update() {
+	if (state == ProjectileState::EXPIRED) {
+		return;
 	}
-	if (phase == P_UNINITED) {
+
+	if (state == ProjectileState::NEW) {
 		Setup();
+		state = ProjectileState::TRAVELLING;
+		return;
 	}
 
 	int pause = core->IsFreezed();
 	if (pause) {
-		return 1;
+		return;
 	}
 
 	const Game *game = core->GetGame();
-	if (game && game->IsTimestopActive() && !(TFlags&PTF_TIMELESS)) {
-		return 1;
+	if (game && game->IsTimestopActive() && !(TFlags & PTF_TIMELESS)) {
+		return;
 	}
 
-	//recreate path if target has moved
-	if(Target) {
-		SetTarget(Target, false);
+	switch (state) {
+		case ProjectileState::TRAVELLING:
+		case ProjectileState::HIT:
+			state = DoStep();
+			break;
+		case ProjectileState::AWAITING_HIT:
+			state = CheckTrigger(Extension->TriggerRadius);
+			break;
+		case ProjectileState::EXPLODING:
+		case ProjectileState::EXPLODING_AGAIN:
+			state = CheckTrigger(Extension->TriggerRadius);
```

Comment:
This assignment is redundant the way it is now. But before what is now `GetNextExplosionState` didn't get called if the trigger got revoked, so I suspect we're actually missing more conditional logic here.

Before:
```c++
            CheckTrigger(Extension->TriggerRadius);
            if (phase == P_EXPLODING1 || phase == P_EXPLODING2) {
                DrawExplosion(viewport, flags);
            }
```

Reply:
Right, I've added a check whether `GetNextExplosionState ` is still applicable after `CheckTrigger`. But the assignment being redundant?

Reply:
Looks fine now and subsequently there's no more redundant assignment (before the first one was unused).